text/0000-build-std.md

```diff
@@ -0,0 +1,3601 @@
+will only additionally recognise `core`, `alloc` and `std` and none of their
+dependencies. Many of Cargo's subcommands will need modification to support
+build-std:
+
+[`cargo add`][cargo-add] will add `core`, `alloc` or `std` explicitly to the
+manifest if invoked with those crate names:
+
+```toml
+[package]
+name = "hello_world"
+version = "0.1.0"
+edition = "2024"
+
+[dependencies]
+std = { builtin = true } # <-- this would be added
+```
+
+[`cargo clean`][cargo-clean] will additionally delete any builds of the standard
+library performed by build-std.
+
+[`cargo fetch`][cargo-fetch] will not fetch the standard library dependencies as
+they are already vendored in the `rust-src` component.
+
+[`cargo info`][cargo-info] will learn how to print information for the built-in
+`std`, `alloc` and `core` dependencies:
+
+```shell-session
+$ cargo info std
+std
+rust standard library
+license: Apache 2.0 + MIT
+rust-version: 1.86.0
+documentation: https://doc.rust-lang.org/1.86.0/std/index.html
+```
+
+```shell-session
+$ cargo info alloc
+alloc
+rust standard library
+license: Apache 2.0 + MIT
+rust-version: 1.86.0
+documentation: https://doc.rust-lang.org/1.86.0/alloc/index.html
+```
+
+```shell-session
+$ cargo info core
+core
+rust standard library
+license: Apache 2.0 + MIT
+rust-version: 1.86.0
+documentation: https://doc.rust-lang.org/1.86.0/core/index.html
+```
+
+[`cargo metadata`][cargo-metadata] will emit `std`, `alloc` and `core`
+dependencies to the metadata emitted by `cargo metadata` (when those crates are
+dependencies). None of the standard library's dependencies will be included.
+`source` would be set to `builtin` and the remaining fields would be set like
+any other dependency.
+
+**TODO:** show example output
+
+[`cargo miri`][cargo-miri] could be re-implemented using build-std to enable a
+`miri` profile and always rebuild. The `miri` profile would be configured in the
+standard library's workspace, setting the flags/options necessary for `miri`.
+
+[`cargo pkgid`][cargo-pkgid] when passed `-p core` would print `builtin#core` as
+the source, likewise with `alloc` and `std`.
+
+[`cargo report`][cargo-report] will not include reports from the standard
+library crates or their dependencies.
+
+[`cargo remove`][cargo-remove] will remove `core`, `alloc` or `std` explicitly
+from the manifest if invoked with those crate names:
+
+```toml
+[package]
+name = "hello_world"
+version = "0.1.0"
+edition = "2024"
+
+[dependencies]
+std = { builtin = true } # <-- this would be removed
+```
+
+[`cargo tree`][cargo-tree] will show `std`, `alloc` and `core` at appropriate
+places in the tree of dependencies. `alloc` will always be shown as a dependency
+of `std`, and `core` a dependency of `alloc`. As opaque dependencies, none of
+the other dependencies of `std`, `alloc` or `core` will be shown. Neither `std`,
+`alloc` or `core` will have a version number.
+
+```shell-session
+$ cargo tree
```

Comment:
I don't think this output is representative - all dependencies would depend on at least core. In addition I believe the vertically first instance of a dependency is expanded and then lower ones are collapsed.

Perhaps clarify implied std dependencies here too.

Reply:
I think the deduplication that `cargo tree` does would mean that `core` would not need to be a part of the tree for every dep. I think the current example is sufficient to illustrate how this would look.